server/src/main/java/org/opensearch/search/aggregations/bucket/filterrewrite/rangecollector/SubAggRangeCollector.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.aggregations.bucket.filterrewrite.rangecollector;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.BitDocIdSet;
+import org.apache.lucene.util.FixedBitSet;
+import org.opensearch.search.aggregations.BucketCollector;
+import org.opensearch.search.aggregations.LeafBucketCollector;
+import org.opensearch.search.aggregations.bucket.filterrewrite.FilterRewriteOptimizationContext;
+import org.opensearch.search.aggregations.bucket.filterrewrite.Ranges;
+
+import java.io.IOException;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import static org.apache.lucene.search.DocIdSetIterator.NO_MORE_DOCS;
+
+/**
+ * Range collector implementation that supports sub-aggregations by collecting doc IDs.
+ */
+public class SubAggRangeCollector extends AbstractRangeCollector {
```

Comment:
IMO, `SubAggRangeCollector` can directly extend from `SimpleRangeCollector` and we can get rid of `AbstractRangeCollector`. Then, it should also delegate to `super.finalizePreviousRange` in `finalizePreviousRange`. This makes sense to me, since `SubAggRangeCollector` is doing all the work of `SimpleRangeCollector` and then some more!

Reply:
Yeah, I think either way is fine.
Good point about SubAggRangeCollector is doing all the work of SimpleRangeCollector and then some more.
I change SubAggRangeCollector to extend SimpleRangeCollector to indicate that.

Will keep AbstractRangeCollector to hold the utility methods, so the actual implementations are cleaner.